shapes.js

```diff
@@ -0,0 +1,348 @@
+  this.interval = 30;
+  setInterval(function() { myState.draw(); }, myState.interval);
+}
+
+CanvasState.prototype.addShape = function(shape) {
+  this.shapes.push(shape);
+  this.valid = false;
+}
+CanvasState.prototype.removeShape = function(shape) {
+  var index = this.shapes.indexOf(shape);
+  this.shapes.splice(index, 1); 
+  this.valid = false;
+}
+
+CanvasState.prototype.clear = function() {
+  this.ctx.clearRect(0, 0, this.width, this.height);
+}
+
+// While draw is called as often as the INTERVAL variable demands,
+// It only ever does something if the canvas gets invalidated by our code
+CanvasState.prototype.draw = function() {
+  // if our state is invalid, redraw and validate!
+  if (!this.valid) {
+    var ctx = this.ctx;
+    var shapes = this.shapes;
+    this.clear();
+    
+    // ** Add stuff you want drawn in the background all the time here **
+    
+    // draw all shapes
+    var l = shapes.length;
+    for (var i = 0; i < l; i++) {
+      var shape = shapes[i];
+      // We can skip the drawing of elements that have moved off the screen:
+      if (shape.x > this.width || shape.y > this.height ||
+          shape.x + shape.w < 0 || shape.y + shape.h < 0) continue;
+      shapes[i].draw(ctx);
+    }
+    
+    // draw selection
+    // right now this is just a stroke along the edge of the selected Shape
+    if (this.selection != null) {
+      ctx.strokeStyle = this.selectionColor;
+      ctx.lineWidth = this.selectionWidth;
+      var mySel = this.selection;
+      ctx.strokeRect(mySel.x,mySel.y,mySel.w,mySel.h);
+
+      /* Handles corners. */
+      ctx.fillStyle="#000000";
+      ctx.fillRect(mySel.x - 2, mySel.y - 2, 5, 5);
+      ctx.fillRect(mySel.x - 3 + mySel.w, mySel.y - 2, 5, 5);
+      ctx.fillRect(mySel.x - 2, mySel.y - 2 + mySel.h, 5, 5);
+      ctx.fillRect(mySel.x - 3 + mySel.w, mySel.y - 2 + mySel.h, 5, 5);
+    }
+    
+    // ** Add stuff you want drawn on top all the time here **
+    
+    this.valid = true;
+  }
+}
+
+
+// Creates an object with x and y defined, set to the mouse position relative to the state's canvas
+// If you wanna be super-correct this can be tricky, we have to worry about padding and borders
+CanvasState.prototype.getMouse = function(e) {
+  var element = this.canvas, offsetX = 0, offsetY = 0, mx, my;
+  
+  // Compute the total offset
+  if (element.offsetParent !== undefined) {
+    do {
+      offsetX += element.offsetLeft;
+      offsetY += element.offsetTop;
+    } while ((element = element.offsetParent));
+  }
+
+  // Add padding and border style widths to offset
+  // Also add the <html> offsets in case there's a position:fixed bar
+  offsetX += this.stylePaddingLeft + this.styleBorderLeft + this.htmlLeft;
+  offsetY += this.stylePaddingTop + this.styleBorderTop + this.htmlTop;
+
+  mx = e.pageX - offsetX;
+  my = e.pageY - offsetY;
+  
+  // We return a simple javascript object (a hash) with x and y defined
+  return {x: mx, y: my};
+}
+
+video_name = "test_vid"
```

Comment:
I forgot to put semicolons when writing my code. Let's use semicolons everywhere even though they're optional. Mind fixing my code for me?